pkg/controller/volume/attachdetach/cache/actual_state_of_world.go

```diff
@@ -96,10 +96,13 @@ type ActualStateOfWorld interface {
 	// nodes, the volume is also deleted.
 	DeleteVolumeNode(volumeName v1.UniqueVolumeName, nodeName types.NodeName)
 
-	// IsVolumeAttachedToNode returns true if the specified volume/node combo exists
-	// in the underlying store indicating the specified volume is attached to
-	// the specified node.
-	IsVolumeAttachedToNode(volumeName v1.UniqueVolumeName, nodeName types.NodeName) bool
+	// GetAttachState returns the attach state for the given volume-node
+	// combination.
+	// Returns AttachStateAttached if the specified volume/node combo exists in
+	// the underlying store indicating the specified volume is attached to the
+	// specified node, AttachStateDetached if the combo does not exist, or
+	// AttachStateUncertain if the attached state is marked as uncertain.
+	GetAttachState(volumeName v1.UniqueVolumeName, nodeName types.NodeName) AttachState
```

Comment:
Why introduce a new more complicated struct here? As far as I can tell callers only use 2 states and none of them appear to be using the new `AttachStateUncertain` state?

Reply:
Between the two call sites, [one of them](https://github.com/kubernetes/kubernetes/pull/94599/commits/d9a629fe3a5c812883a8c22744d3571facaceb44#diff-359fd18b3708873b0dc50ba24ba0b056f9401e66d73351bea46ee30f95b5b4b9R465)  matches exactly on attach while [the other](https://github.com/kubernetes/kubernetes/pull/94599/commits/d9a629fe3a5c812883a8c22744d3571facaceb44#diff-e76cff1d139f1df6d8c37cea61b19aa8caea6b103d64e58a084f2d81c5bad149R173) matches exactly on detach. Uncertain is not an acceptable state for either caller. Before, "uncertain" and "detached" states are both "false"

A type with 3 values is IMO more explicit than a type with 2 values, and simplifies [this logic](https://github.com/kubernetes/kubernetes/pull/94599/commits/d9a629fe3a5c812883a8c22744d3571facaceb44#diff-e76cff1d139f1df6d8c37cea61b19aa8caea6b103d64e58a084f2d81c5bad149L172). But strictly speaking it's not necessary functionally